Multi-class, multi-output RandomForestClassifier in scikit-learn produces error

Reproducer:
```python
import numpy as np
import treelite
from sklearn.datasets import make_classification
from sklearn.ensemble import RandomForestClassifier
from sklearn.utils import shuffle

X, y1 = make_classification(
    n_samples=10, n_features=100, n_informative=30, n_classes=3, random_state=1
)
y2 = shuffle(y1, random_state=1)
y3 = shuffle(y1, random_state=2)
Y = np.vstack((y1, y2, y3)).T
print(X.shape, Y.shape)
n_samples, n_features = X.shape
n_outputs = Y.shape[1]
n_classes = 3
forest = RandomForestClassifier(random_state=1)
forest.fit(X, Y)

model = treelite.sklearn.import_model(forest)
```

Error:
```
  File "treelite/sklearn/importer.py", line 46, in add
    array.shape == expected_shape
AssertionError: Expected shape: (13, 3, [3, 3, 3]), Got shape (13, 3, 3)
```

--- a/.github/workflows/windows-wheel-builder.yml
+++ b/.github/workflows/windows-wheel-builder.yml
@@ -36,10 +36,6 @@ jobs:
         cache-env: true
         environment-name: dev
         environment-file: ops/conda_env/dev.yml
-    - name: Display Conda env
-      run: |
-        micromamba info
-        micromamba list
     - name: Build wheel
       run: |
         call ops/build-windows.bat

--- a/ops/test-win-python-wheel.bat
+++ b/ops/test-win-python-wheel.bat
@@ -6,7 +6,7 @@ setlocal enabledelayedexpansion
 python tests\ci_build\rename_whl.py python\dist %COMMIT_ID% win_amd64
 if %errorlevel% neq 0 exit /b %errorlevel%
 for /R %%i in (python\\dist\\*.whl) DO (
-  python -m pip install "%%i"
+  python -m pip install --force-reinstall "%%i"
   if !errorlevel! neq 0 exit /b !errorlevel!
 )
 

--- a/python/treelite/sklearn/importer.py
+++ b/python/treelite/sklearn/importer.py
@@ -139,10 +139,14 @@ def import_model(sklearn_model):
         )
     elif isinstance(sklearn_model, (RandomForestC, ExtraTreesC)):
         # pylint: disable=C3001
+        try:  # multi-target
+            max_n_classes = max(sklearn_model.n_classes_)
+        except TypeError:  # single-target
+            max_n_classes = sklearn_model.n_classes_
         leaf_value_expected_shape = lambda node_count: (  # noqa: E731
             node_count,
             sklearn_model.n_outputs_,
-            sklearn_model.n_classes_,
+            max_n_classes,
         )
     elif isinstance(
         sklearn_model, (GradientBoostingR, GradientBoostingC, IsolationForest)

--- a/src/model_loader/sklearn.cc
+++ b/src/model_loader/sklearn.cc
@@ -86,7 +86,7 @@ class RandomForestClassifierMixIn {
     TREELITE_CHECK_GT(max_num_class_, 0)
         << "max_num_class not yet initialized. Was HandleMetadata() called?";
     std::vector<double> leafvec(&value[tree_id][node_id * n_targets_ * max_num_class_],
-        &value[tree_id][(node_id + 1) * max_num_class_]);
+        &value[tree_id][(node_id + 1) * n_targets_ * max_num_class_]);
     // Compute the probability distribution over label classes
     double const norm_factor = std::accumulate(leafvec.begin(), leafvec.end(), 0.0);
     std::for_each(leafvec.begin(), leafvec.end(), [norm_factor](double& e) { e /= norm_factor; });

--- a/src/model_loader/sklearn.cc
+++ b/src/model_loader/sklearn.cc
@@ -14,6 +14,7 @@
 
 #include <algorithm>
 #include <cstdint>
+#include <experimental/mdspan>
 #include <limits>
 #include <memory>
 #include <numeric>
@@ -22,6 +23,11 @@
 namespace treelite::model_loader::sklearn {
 
 namespace detail {
+
+namespace stdex = std::experimental;
+// Multidimensional array views. Use row-major (C) layout
+template <typename ElemT>
+using Array2DView = stdex::mdspan<ElemT, stdex::dextents<std::uint64_t, 2>, stdex::layout_right>;
 
 class RandomForestRegressorMixIn {
  public:
@@ -80,7 +86,7 @@ class RandomForestClassifierMixIn {
   }
 
   void HandleLeafNode(model_builder::ModelBuilder& builder, int tree_id, int node_id,
-      double const** value, [[maybe_unused]] int const* n_classes) const {
+      double const** value, std::int32_t const* n_classes) const {
     TREELITE_CHECK_GT(n_targets_, 0)
         << "n_targets not yet initialized. Was HandleMetadata() called?";
     TREELITE_CHECK_GT(max_num_class_, 0)
@@ -88,8 +94,16 @@ class RandomForestClassifierMixIn {
     std::vector<double> leafvec(&value[tree_id][node_id * n_targets_ * max_num_class_],
         &value[tree_id][(node_id + 1) * n_targets_ * max_num_class_]);
     // Compute the probability distribution over label classes
-    double const norm_factor = std::accumulate(leafvec.begin(), leafvec.end(), 0.0);
-    std::for_each(leafvec.begin(), leafvec.end(), [norm_factor](double& e) { e /= norm_factor; });
+    auto leaf_view = Array2DView<double>(leafvec.data(), n_targets_, max_num_class_);
+    for (int target_id = 0; target_id < n_targets_; ++target_id) {
+      double norm_factor = 0.0;
+      for (std::int32_t class_id = 0; class_id < n_classes[target_id]; ++class_id) {
+        norm_factor += leaf_view(target_id, class_id);
+      }
+      for (std::int32_t class_id = 0; class_id < n_classes[target_id]; ++class_id) {
+        leaf_view(target_id, class_id) /= norm_factor;
+      }
+    }
     builder.LeafVector(leafvec);
   }
 
@@ -116,7 +130,7 @@ class IsolationForestMixIn {
   }
 
   void HandleLeafNode(model_builder::ModelBuilder& builder, int tree_id, int node_id,
-      double const** value, [[maybe_unused]] int const* n_classes) const {
+      double const** value, [[maybe_unused]] std::int32_t const* n_classes) const {
     builder.LeafScalar(value[tree_id][node_id]);
   }
 

--- a/tests/python/test_sklearn_integration.py
+++ b/tests/python/test_sklearn_integration.py
@@ -17,6 +17,7 @@ from .hypothesis_util import (
 from .util import has_pandas, to_categorical
 
 try:
+    from sklearn.datasets import make_classification
     from sklearn.dummy import DummyClassifier, DummyRegressor
     from sklearn.ensemble import (
         ExtraTreesClassifier,
@@ -29,6 +30,7 @@ try:
         RandomForestClassifier,
         RandomForestRegressor,
     )
+    from sklearn.utils import shuffle
 except ImportError:
     # Skip this test suite if scikit-learn is not installed
     pytest.skip("scikit-learn not installed; skipping", allow_module_level=True)
@@ -56,7 +58,7 @@ def test_skl_regressor(clazz, n_estimators, callback):
     else:
         n_targets = callback.draw(just(1))
     X, y = callback.draw(standard_regression_datasets(n_targets=just(n_targets)))
-    kwargs = {"max_depth": 3, "random_state": 0}
+    kwargs = {"max_depth": 8, "random_state": 0}
     if clazz == HistGradientBoostingRegressor:
         kwargs["max_iter"] = n_estimators
     else:
@@ -101,7 +103,7 @@ def test_skl_classifier(clazz, dataset, n_estimators, callback):
     """Scikit-learn binary classifier"""
     X, y = dataset
     n_classes = len(np.unique(y))
-    kwargs = {"max_depth": 3, "random_state": 0}
+    kwargs = {"max_depth": 8, "random_state": 0}
     if clazz == HistGradientBoostingClassifier:
         kwargs["max_iter"] = n_estimators
     else:
@@ -132,6 +134,36 @@ def test_skl_classifier(clazz, dataset, n_estimators, callback):
         and n_classes == 2
     ):
         expected_prob = expected_prob[:, :, 1:]
+    np.testing.assert_almost_equal(out_prob, expected_prob, decimal=5)
+
+
+@given(
+    n_classes=integers(min_value=3, max_value=5),
+    n_estimators=integers(min_value=3, max_value=10),
+)
+@settings(**standard_settings())
+def test_skl_multitarget_multiclass_rf(n_classes, n_estimators):
+    """Scikit-learn RF classifier with multiple outputs and classes"""
+    X, y1 = make_classification(
+        n_samples=1000,
+        n_features=100,
+        n_informative=30,
+        n_classes=n_classes,
+        random_state=0,
+    )
+    y2 = shuffle(y1, random_state=1)
+    y3 = shuffle(y1, random_state=2)
+
+    y = np.vstack((y1, y2, y3)).T
+
+    clf = RandomForestClassifier(
+        max_depth=8, n_estimators=n_estimators, n_jobs=-1, random_state=4
+    )
+    clf.fit(X, y)
+
+    tl_model = treelite.sklearn.import_model(clf)
+    out_prob = treelite.gtil.predict(tl_model, X)
+    expected_prob = clf.predict_proba(X)
     np.testing.assert_almost_equal(out_prob, expected_prob, decimal=5)
 
 
@@ -188,7 +220,7 @@ def test_skl_hist_gradient_boosting_with_categorical(
         max_iter=num_boost_round,
         categorical_features=categorical_features,
         early_stopping=False,
-        max_depth=1,
+        max_depth=8,
     )
     clf.fit(df, y)
 

--- a/tests/python/test_sklearn_integration.py
+++ b/tests/python/test_sklearn_integration.py
@@ -163,7 +163,7 @@ def test_skl_multitarget_multiclass_rf(n_classes, n_estimators):
 
     tl_model = treelite.sklearn.import_model(clf)
     out_prob = treelite.gtil.predict(tl_model, X)
-    expected_prob = clf.predict_proba(X)
+    expected_prob = np.transpose(clf.predict_proba(X), axes=(1, 0, 2))
     np.testing.assert_almost_equal(out_prob, expected_prob, decimal=5)
 
 

